File: backend/app/schemas/cred.py
from datetime import date

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.patient import PatientSummary
from app.schemas.types import LocalDatetime


class CredEntryWrite(BaseModel):
    """Registro de una prestación aplicada."""

    item_code: str = Field(min_length=2, max_length=32)
    performed_on: date
    result: str | None = Field(default=None, max_length=160)
    notes: str | None = Field(default=None, max_length=1000)
    practitioner_id: int | None = None

    @field_validator("result", "notes", mode="before")
    @classmethod
    def _blank_to_none(cls, value: object) -> object:
        if isinstance(value, str):
            cleaned = " ".join(value.split())
            return cleaned or None
        return value


class CredEntryCreate(CredEntryWrite):
    patient_id: int


class CredEntryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    patient_id: int
    item_code: str
    sequence: int
    performed_on: date
    result: str | None
    notes: str | None
    practitioner_id: int | None
    practitioner_name: str | None
    created_at: LocalDatetime


class CredSlot(BaseModel):
    """Una casilla del carné: lo que toca y, si ya se dio, cuándo."""

    item_code: str
    group: str
    label: str
    dose: str
    due_month: int | None
    status: str
    status_label: str
    records_result: bool
    entries: list[CredEntryRead] = Field(default_factory=list)


class CredSection(BaseModel):
    """Una de las rejillas del carné."""

    kind: str
    label: str
    slots: list[CredSlot]
    applied: int
    overdue: int


class CredCard(BaseModel):
    """Carné completo de un niño: el calendario cruzado con lo aplicado."""

    patient: PatientSummary
    age_months: int | None
    age_label: str
    in_program: bool
    sections: list[CredSection]
    applied: int
    overdue: int
    pending: int


class CredSheet(BaseModel):
    """Carné listo para imprimir y entregar."""

    title: str
    number: str
    html: str


class CredCatalogItem(BaseModel):
    """Prestación ofrecida al registrar, ya marcada si no admite repetición."""

    code: str
    kind: str
    kind_label: str
    group: str
    label: str
    dose: str
    due_month: int | None
    records_result: bool
    repeatable: bool
    already_applied: bool

File: frontend/src/types/cred.ts
/** Estados de una casilla del carné, tal como los calcula el servidor. */
export const CRED_STATUSES = ['APLICADA', 'PENDIENTE', 'ATRASADA', 'FUTURA'] as const
export type CredStatus = (typeof CRED_STATUSES)[number]

export const CRED_KINDS = [
  'VACUNA',
  'CONTROL',
  'TAMIZAJE',
  'SUPLEMENTO',
  'PRESTACION',
] as const
export type CredKind = (typeof CRED_KINDS)[number]

export interface CredEntry {
  id: number
  patient_id: number
  item_code: string
  sequence: number
  performed_on: string
  result: string | null
  notes: string | null
  practitioner_id: number | null
  practitioner_name: string | null
  created_at: string
}

/** Una casilla: lo que toca a esa edad y, si ya se dio, cuándo. */
export interface CredSlot {
  item_code: string
  group: string
  label: string
  dose: string
  due_month: number | null
  status: CredStatus
  status_label: string
  records_result: boolean
  entries: CredEntry[]
}

export interface CredSection {
  kind: CredKind
  label: string
  slots: CredSlot[]
  applied: number
  overdue: number
}

export interface CredCard {
  patient: {
    id: number
    public_id: string
    history_number: string
    full_name: string
    age: number | null
    sex: string | null
  }
  age_months: number | null
  age_label: string
  in_program: boolean
  sections: CredSection[]
  applied: number
  overdue: number
  pending: number
}

export interface CredCatalogItem {
  code: string
  kind: CredKind
  kind_label: string
  group: string
  label: string
  dose: string
  due_month: number | null
  records_result: boolean
  repeatable: boolean
  already_applied: boolean
}

export interface CredSheet {
  title: string
  number: string
  html: string
}

export interface CredEntryPayload {
  item_code: string
  performed_on: string
  result: string | null
  notes: string | null
  practitioner_id: number | null
}

export interface CredEntryCreatePayload extends CredEntryPayload {
  patient_id: number
}
